sync/src/synchronizer/block_fetcher.rs

```diff
@@ -123,26 +123,42 @@ impl BlockFetcher {
                     break;
                 }
 
-                let to_fetch = self
-                    .snapshot
-                    .get_ancestor(&best_known_header.hash(), index_height)?;
-
-                // NOTE: Filtering `BLOCK_STORED` but not `BLOCK_RECEIVED`, is for avoiding
-                // stopping synchronization even when orphan_pool maintains dirty items by bugs.
-                if self
-                    .snapshot
-                    .contains_block_status(&to_fetch.hash(), BlockStatus::BLOCK_STORED)
-                {
-                    continue;
-                }
-
-                if inflight.insert(self.peer, to_fetch.hash()) {
-                    trace!(
-                        "[Synchronizer] inflight insert {:?}------------{}",
-                        to_fetch.number(),
-                        to_fetch.hash(),
-                    );
-                    fetch.push(to_fetch.hash());
+                // Try to get [from, to) headers, return the corresponding headers in reversed order.
+                let from = index_height;
+                let to = from + (count - fetch.len()) as u64;
+                let candidates = {
+                    let mut candidates = Vec::with_capacity((to - from) as usize);
+                    let mut parent_hash = self
+                        .snapshot
+                        .get_ancestor(&best_known_header.hash(), to)?
```

Comment:
The block number `to` may overflow.

Reply:
why `to` may overflow?

`to = from + (count - fetch.len())`, `from <= tip_number`, `count - fetch.len() < 16`, so I think `to` is safe.

Reply:
Here `to = from + 16`, assume `from = 0, fetch.len() = 0, to = 16`,  in Rust `0..16` means [0, 16), there are 16 items, if 16 included,  there are 17 items. There is possible `to > best_known_header.number()`.

Reply:
Thanks. Already fixed now.